In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.ensemble import IsolationForest
from sklearn import set_config
set_config(transform_output='pandas')

In [2]:
df = pd.read_csv("Data/credit_card.csv")
df.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0  0.098698  0.363787  0.090794  ...  0.251412 -0.018307  0.277838 -0.110474   
1  0.085102 -0.255425 -0.166974  ... -0.069083 -0.225775 -0.638672  0.101288   
2  0.247676 -1.514654  0.207643  ...  0.524980  0.247998  0.771679  0.909412   
3  0.377436 -1.387024 -0.054952  ... -0.208038 -0.108300  0.005274 -0.190321   
4 -0.270533  0.817739  0.753074  ...  0.408542 -0.009431  0.798278 -0.137458   

        V24       V25       V26       V27       V28  Amount  
0  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62  
1 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69  
2 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66  
3 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50  
4  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99  

[5 rows x 29 columns]

## About 0.4% of card transations are fraudulent. We need to detect these anomalies

# KMeans

In [3]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
# We need to scale our data for KMeans
scaler = StandardScaler()
# Fit & transform data.
scaled_df = scaler.fit_transform(df)

In [4]:
# Apply K-Means algorithm
kmeans = KMeans(n_clusters=3, n_init=10, random_state = 42)
kmeans.fit(scaled_df)

KMeans(n_clusters=3, n_init=10, random_state=42)

In [5]:
# Making a dataframe copy of X for saving clusters 
km = scaled_df.copy()
km['cluster'] = kmeans.predict(scaled_df)
km.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -0.726092 -0.270865  1.389980  0.762227 -0.236899  0.264052  0.285346   
1  0.949810 -0.013020 -0.647336  0.110710  0.102482 -0.155239 -0.009963   
2 -0.725138 -1.235035  0.733929  0.062810 -0.377373  1.293991  0.797185   
3 -0.467623 -0.356408  0.750937 -0.808030  0.042565  0.868121  0.283501   
4 -0.593701  0.452248  0.540940  0.079101 -0.295578 -0.018016  0.613062   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.131083 -0.384443  0.286120  ...  0.043066  0.689788 -0.163478  0.083851   
1  0.120253 -0.915428  0.051569  ... -0.185510 -0.760673  0.268858 -0.603364   
2  0.249752 -1.995239  0.392444  ...  0.336466  1.471338  1.918734 -1.193707   
3  0.353112 -1.885794  0.153501  ... -0.056083  0.258431 -0.326493 -2.015264   
4 -0.163028  0.004829  0.888749  ...  0.052846  1.513434 -0.218569  0.209441   

        V25       V26       V27       V28    Amount  cluster  
0  0.099198 -0.532722  0.312929 -0.090009  0.459472        0  
1  0.188884  0.033761 -0.039088  0.043500 -0.333835        0  
2 -0.959876 -0.442774 -0.153601 -0.234422  1.696110        0  
3  1.303733 -0.591732  0.137995  0.217896  0.318445        0  
4 -0.677493  0.710640  0.524976  0.791442  0.029532        0  

[5 rows x 30 columns]

In [6]:
# Saving the cluster centers as a dataframe for visibility
cluster_centers = pd.DataFrame(kmeans.cluster_centers_, columns=scaled_df.columns)
cluster_centers

V1        V2        V3        V4        V5        V6        V7  \
0 -0.061196  0.053226 -0.050620 -0.094205 -0.118248 -0.244715  0.185121   
1  0.032499 -0.008109  0.190288  0.151972 -0.140992 -0.145277 -0.054969   
2  0.062160 -0.223727 -1.324018 -0.762400  1.835461  2.565302 -0.552852   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0 -0.035698 -0.680309  0.108181  ...  0.041215  0.084629  0.142682 -0.045672   
1 -0.042637  0.463768 -0.088042  ... -0.056608 -0.044251 -0.009889  0.028587   
2  0.554719 -0.171603  0.147610  ...  0.250390 -0.091785 -0.699499  0.009924   

        V24       V25       V26       V27       V28    Amount  
0 -0.095136 -0.013498 -0.189528  0.098632  0.013024  0.007409  
1 -0.125670 -0.042935  0.163101 -0.067390 -0.007712 -0.004050  
2  1.579702  0.435446 -0.333151  0.026162 -0.006537 -0.006551  

[3 rows x 29 columns]

In [7]:
from scipy.spatial.distance import cdist
# Calculate distance to each cluster center
distances = cdist(scaled_df, kmeans.cluster_centers_, 'euclidean')

In [8]:
# Saving distances as a dataframe for convenience
cluter_cols = [f"Distance (Cluster {c})" for  c in range(len(kmeans.cluster_centers_))]
distance_df = pd.DataFrame(distances, columns = cluter_cols)
distance_df.head(3)

Distance (Cluster 0)  Distance (Cluster 1)  Distance (Cluster 2)
0              2.804682              4.393264              5.857215
1              2.426430              3.686087              5.075081
2              6.542622              7.337378              8.656002

In [9]:
# Get the minimum distance to any cluster for each point
min_distances = np.min(distances, axis=1)
# Set a threshold based on our known parameter
threshold = np.percentile(min_distances, 99.6)
threshold

20.909887868001256

In [10]:
# Identify anomalies where the distance to closest cluster center is above the threshold
filter_anomalies = min_distances > threshold
# how many were found?
filter_anomalies.sum()

40

In [20]:
filter_anomalies

array([False, False, False, ..., False, False, False])

### There are forty fraudulent cases in our dataset

In [12]:
# Getting the row indices of the anomalies
idx_anomalies = df[filter_anomalies].index
# Slicing the anomalies from X (for demonstration)
anoms_km = df.iloc[idx_anomalies]
anoms_km

V1         V2         V3         V4         V5         V6  \
159   -6.093248 -12.114213  -5.694973   3.294389  -1.413792   4.776000   
1376  -3.936794  -3.670519  -1.453820   2.299750  -9.749440   6.456410   
1619 -11.140706  -9.612726 -12.389545   6.013346 -32.092129  21.393069   
2156  -5.849226 -11.754975  -3.333303   3.673246  -4.459583   0.659098   
2212  -5.149638   4.261740  -1.043112  -2.872586  -0.970562   2.037714   
2439 -12.168192 -15.732974  -0.376474   3.792613  10.658654  -7.465603   
2594  -2.574940  -2.669331  -3.080133   2.043551 -11.416881   7.885721   
2654  -7.546256   3.051033  -0.524684   1.005228   0.040486   2.751676   
2756 -10.780565  -9.940571   1.703409   3.094958   7.929119  -4.032777   
2911  -4.344234   4.268511  -1.580163   0.272536  -1.419244   2.603639   
2914  -5.390190   5.141127  -2.456760  -1.433851  -2.132299   2.693505   
2917  -5.948482   5.573239  -2.932061  -1.450307  -2.521964   2.641657   
2923  -6.200114   5.025406  -2.742492  -0.940903  -6.656259   5.432294   
3443  -5.910028   5.301319  -0.041838   1.537238  -1.059957   3.614102   
4779  -5.043472   4.015914   1.999942  -3.470182   1.755487  -0.376547   
5303  -8.733429  -5.681953   2.253879   3.930311  -0.916149   4.355297   
5412  -8.053149   7.045990  -1.102994  -0.870690  -1.360528   4.754590   
5413  -8.617888   7.467017  -1.565033  -0.884684  -1.699891   4.669832   
5529  -4.714214   3.917665  -1.304351   0.740001   0.399374   3.014229   
5674  -8.382273  -9.039038   0.499895  -0.574757   4.242717  -2.990519   
5704  -5.522861   5.065940  -2.422346  -0.739464  -1.717152   4.098583   
5764  -9.092667  -7.878057  -0.080916  -1.492171   1.206717  -1.222811   
6489  -9.175447   7.906022  -2.068334  -0.901683  -2.178911   4.675277   
6643  -6.571336 -11.943892  -4.246322   6.825816  -2.896624   2.706661   
6672 -23.066842 -25.640527  -3.080313   4.866932   6.901397  -4.074335   
7322  -3.584791  -2.855326   0.271564   4.772815   0.988323   0.352907   
7338  -9.507807  -6.332182  -2.347929   5.909158  -1.946121   2.347826   
7470  -6.487286 -13.630449  -3.190735   2.890879  -5.115535   2.393873   
7596  -9.791064   8.261750  -2.524941  -0.896418  -2.430637   4.530167   
7597  -6.057491   5.243408  -1.809137   0.666109  -3.233775   5.409330   
8124  -2.125490   5.973556 -11.034727   9.007147  -1.689451  -2.854415   
8163  -1.426623   4.141986  -9.804103   6.666273  -4.749527  -2.073129   
8437 -27.670569 -34.607649   2.243079  10.463020  34.099309 -23.496714   
8442  -3.891192   7.098916 -11.426467   8.607557  -2.065706  -2.985288   
8856  -2.589617   7.016714 -13.705407  10.343228  -2.954461  -3.055116   
8939 -10.884788 -13.449684  -1.566937   1.034831   3.528629  -3.459355   
8999  -2.880042   5.225442 -11.063330   6.689951  -5.759924  -2.244031   
9071  -5.454362   8.287421 -12.752811   8.594342  -3.106002  -3.179949   
9304  -4.153014   8.204797 -15.031714  10.330100  -3.994426  -3.250013   
9326  -4.710529   8.636214 -15.496222  10.313349  -4.351341  -3.322689   

             V7         V8        V9        V10  ...        V20        V21  \
159    4.808426  -0.228197 -0.525896  -1.724899  ...   7.744222   2.228823   
1376  10.784088  -2.160016 -0.238116  -1.004398  ...  -2.203676  -0.979104   
1619  34.303177  -7.520784 -1.925732  -2.636622  ... -11.748689  -4.709977   
2156   4.527224  -0.985512  0.237794  -1.854230  ...   7.314347   2.160870   
2212  -4.472909 -14.892657  1.713550   0.090612  ...  -3.235393  14.718212   
2439  -6.907038   1.573722  0.058164  -0.421881  ...   4.469095   1.660209   
2594  12.219249  -1.828209 -1.475925  -1.965803  ...   0.828357  -0.115580   
2654  -6.267406 -19.176657  1.049976   0.521925  ...  -3.785808   9.639417   
2756  -8.945496  -5.099389  2.297646  -0.234243  ...  -4.829446  -4.399837   
2911 -11.164794 -23.632502 -1.849347  -5.217814  ...   5.851381 -11.273890   
2914  -9.704282 -23.179640 -1.747019  -6.042641  ...   5.580590 -11.102491   
2917 -10.045540 -22.746370 -1.726559  -5.9

In [16]:
idx_anomalies

Int64Index([ 159, 1376, 1619, 2156, 2212, 2439, 2594, 2654, 2756, 2911, 2914,
            2917, 2923, 3443, 4779, 5303, 5412, 5413, 5529, 5674, 5704, 5764,
            6489, 6643, 6672, 7322, 7338, 7470, 7596, 7597, 8124, 8163, 8437,
            8442, 8856, 8939, 8999, 9071, 9304, 9326],
           dtype='int64')

### Above we see the forty anomalies that are mostly likely to be fraudulent

# Isolation Forest

In [13]:
# Instantiate the model with a contaimination of 0.004 (we will identify 0.4% as anamolous)
iso = IsolationForest(contamination=0.004, random_state = 42)
# fit the model using .values to avoid a warning
iso.fit(df.values)

IsolationForest(contamination=0.004, random_state=42)

In [14]:
# Obtain results from the model
predictions = iso.predict(df.values)
predictions[:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [15]:
# Change the labels to match our columns from the kmeans dataframe
# Not anamolies
predictions[predictions ==1] = 0
# Anomalies
predictions[predictions ==-1] = 1
# Preview new labels
predictions[:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [17]:
# Add isolation forest predictions to the dataframe
df['iso_anomalies'] = predictions

In [19]:
filter_iso = df['iso_anomalies'] == 1
filter_iso

0       False
1       False
2       False
3       False
4       False
        ...  
9995    False
9996    False
9997    False
9998    False
9999    False
Name: iso_anomalies, Length: 10000, dtype: bool

In [22]:
iso_anomalies = df[filter_iso].index
iso_anomalies

Int64Index([ 159, 1619, 2156, 2756, 2858, 2914, 2917, 2923, 5303, 5412, 5413,
            5704, 6311, 6489, 6581, 6595, 6634, 6643, 6672, 6757, 6761, 6798,
            6829, 7338, 7470, 7596, 7597, 8124, 8163, 8437, 8442, 8627, 8645,
            8667, 8670, 8856, 8999, 9071, 9304, 9326],
           dtype='int64')

In [23]:
# Make a list of anomolies identified in both methods
both = [a for a in iso_anomalies if a in idx_anomalies]
both

[159,
 1619,
 2156,
 2756,
 2914,
 2917,
 2923,
 5303,
 5412,
 5413,
 5704,
 6489,
 6643,
 6672,
 7338,
 7470,
 7596,
 7597,
 8124,
 8163,
 8437,
 8442,
 8856,
 8999,
 9071,
 9304,
 9326]

In [26]:
# How many anomalies overlap?
len(both)

27

### Twenty-seven anomalies occurred across both models

In [29]:
# Calculating percentage of overlap
p = (27/40)*100
p

67.5

### With 10000 cases in this df and 0.04% being fraudulent, about 40 anomalies should be present. If so, there's a 67.5% overlap between KMeans and Isolation Forest